Updated matches backend

Astraeus- opened the pull request:
Fixes #85 #86  
Removed the match round from update match. I don't think this needs to editable after a match has been created. 

Disabled updating of the play and schedule deadlines from the update match backend.

Renamed fields to be more clear, added more intuitive ordering and made create/update match consistent.

Display teams participating in a match and filtered the winner selection based on the these participating teams, instead of the old massive drop down.

rogusdev commented:
This looks great!  It does have side effects that were not at all expected to me -- that I further like,m but want to understand where they come from: Caster and Timeline have now moved to separate tabs, and the tab design has changed.

Old:
![edit_match_old](https://user-images.githubusercontent.com/1178706/54879515-2fce4000-4e10-11e9-8726-891bdc27ca36.png)

New:
![edit_match_new](https://user-images.githubusercontent.com/1178706/54879516-32309a00-4e10-11e9-9e79-bafaaa5812c1.png)

How did those changes happen?  Very happy with them, just want to understand the system.

Astraeus- commented:
That's part of my backend refactor in 48540d1482f7c174c473b682673266a62d57591d specifically for matches: `plugins/rikki/heroeslounge/models/match/fields.yaml` and `plugins/rikki/heroeslounge/models/match/fields_on_create.yaml`

rogusdev commented:
Doh... I had not rebased my branch on latest upstream master... mb, thanks!

Astraeus- commented:
Also fixes #15 by adding a custom dropdown instead of purely the relation to the match model. @Fabian-Sommer when you deploy there are some divisions that belong to a playoff that doesn't exist. (playoff_id: 12). These need to be deleted or I need to account for divisions that used to belong to a playoff that does not exist anymore.

rogusdev commented on plugins/rikki/heroeslounge/models/Match.php:
```diff
@@ -82,6 +83,32 @@ class Match extends Model
+        $list = [];
+
+        foreach ($divisions as $division) {
+            if (!is_null($division->season_id)) {
```
Should this logic live on the division model so that it can be reused elsewhere?  This kind of naming seems like it is not unique to this dropdown, potentially.

Astraeus- replied:
I'm not sure if it's worth moving this to the division model. However I did find that there's already a method `getLongTitleAttribute()` on the division model. It would probably be worth updating that logic and using it to create the long division title for each division.

rogusdev replied:
oh yeah you should definitely modify that method because it is almost exactly the same as this logic already, and it does not appear to be used anywhere...  (So changing it is totally safe and would be perfect here)

Astraeus- commented:
Added the logic to the division model instead of the match model. Slightly refactored the existing `longTitle` logic and used that instead of my own to populate the division drop down.

rogusdev commented on plugins/rikki/heroeslounge/models/Match.php:
```diff
@@ -9,6 +9,7 @@
 use Rikki\Heroeslounge\Models\Season as Season;
 use Rikki\Heroeslounge\Models\Playoff;
 use Rikki\Heroeslounge\Models\Team;
+use Rikki\Heroeslounge\Models\Division as Division;
```
Is this `as` not redundant since that's what it is already? (Same for season above)

Fabian-Sommer replied:
It is

Fabian-Sommer commented on plugins/rikki/heroeslounge/models/match/fields.yaml:
```diff
@@ -1,40 +1,43 @@
-        round:
-            label: 'Round'
-            span: left
-            type: number
```
Please allow round to be edited after creation - people sometimes make mistakes, this gives them the option to fix their mistake.

Fabian-Sommer commented on plugins/rikki/heroeslounge/models/Match.php:
```diff
@@ -82,6 +83,16 @@ class Match extends Model
         ]
     ];
 
+    public function listWinnerOptions($fieldname, $value, $formData)
```
Ideally, this should be part of the controller since it is very specific for the use in backend. Not sure if it can be moved easily. If it stays in the model, please rename it to listParticipatingTeamsForBackend or something like this

Astraeus- replied:
See below response, as it's somewhat similar.

Fabian-Sommer commented on plugins/rikki/heroeslounge/models/Match.php:
```diff
@@ -82,6 +83,16 @@ class Match extends Model
+    public function divisionsLongTitleDropdown()
+    {
+        return Division::listDivisionsWithLongTitle();
+    }
```
If at all possible, this should not be part of the model, but of the controller - This has nothing to do with the match at all!

Astraeus- replied:
According to the [OctoberCMS documentation](https://octobercms.com/docs/backend/forms#form-fields) for form fields, the field definitions should be defined on the model class file, unless I'm misunderstanding their documentation.

For this specific case, because it's a form for creating a match, I'm quite certain that the accompanying `fields_on_create.YAML` file can only reference the match class to populate it's data.

Fabian-Sommer replied:
Could you point me to the part of the docs that make you think so? Unless my understanding of MVC is entirely off this is not something that goes in the model.

Also, it should be static if October is really so inflexible that it HAS to be there

Astraeus- replied:
Moved to the match controller

Fabian-Sommer commented on plugins/rikki/heroeslounge/models/Division.php:
```diff
@@ -85,12 +85,23 @@ public function getLongTitleAttribute()
+        foreach ($divisions as $division) {
+            $list[$division->id] = $division->longTitle;
+        }
+        return $list;
```
This can be done with a single line using the collection and map

Astraeus- replied:
I attempted to use `map`, but I'm not quite sure how to get the output that I need. The following is the closest I got, but I require an array of key value pairs of `$divison->id => $division->longTitle` for attaching the correct division to a match.

```php
Division::all()->map(function ($division) {
    return [$division->id => $division->longTitle];
})>flatten()->toArray();

Output:
  0 => 'Season 4 - Division 1',
  1 => 'Season 4 - Division 2',
  2 => 'Season 4 - Division 3',
...

Desired output:
  1 => 'Season 4 - Division 1',
  2 => 'Season 4 - Division 2',
  3 => 'Season 4 - Division 3',
...
```

Advice how to accomplish this is appreciated.

Fabian-Sommer replied:
I think the problem in your current solution is that you at first create small arrays and then use flatten which causes the keys to be enumerated. Maybe try 
`Division::all()->map(function ($division) {
    return $division->id => $division->longTitle;
})->toArray();`

rogusdev replied:
Better languages can solve this problem the way you describe, but php is far from that good.  As written, this really is the best way in php.  Here is some discussion of the options:
https://stackoverflow.com/questions/11563119/how-to-convert-an-array-of-arrays-or-objects-to-an-associative-array

rogusdev replied:
Oh although I dug a bit deeper and it turns out you can do it with array map if you jump through some extra hoops:
https://ryanwinchester.ca/posts/php-array-map-with-keys

Look at example #3 from the official docs:
https://www.php.net/manual/en/function.array-map.php

Fabian-Sommer replied:
Did you have a look at https://octobercms.com/docs/services/collections
Which is what October uses in many places instead of a bare PHP array

Astraeus- replied:
Figured it out by using `collection->keyBy()` before `collection->map()`.

Astraeus- commented:
Re-added the ability to update a match round. Also moved the `divisionlongtitle` dropdown off of the match model to the match controller model, making certain to only render it on the `create` form. I added the `round` field there as well, because I wasn't sure how else to maintain the field order.

I couldn't figure out how to reference the selected match model from the controller when updating, so I wasn't able to move `listParticipatingTeamsForBackend` to the controller.

Astraeus- commented on plugins/rikki/heroeslounge/controllers/Match.php:
```diff
@@ -24,6 +25,25 @@ public function __construct()
+                'type' => 'number'
+            ]
+            ]);
+    }
```
Yeah that's how I understood how logic needs to be divided when using `yaml` forms from the OctoberCMS documentation. It made me feel like the best practice is to define most of the logic on the model class instead of the controller, which makes the current implementation look like a bit of a "hack" to me.

Alternatively we could no longer use `yaml` forms, but I think their ease of use and the required rewrite outweighs doing so.

Fabian-Sommer replied:
I really disliked having a static method in Match that just returns the result of a static method in Division. With this, it convolutes a bit where the fields are defined, but I think that is ok considering the alternative.
However, I think that the Round can be moved back to the .yaml - at least I see no reason why it has to be here.

Side note: Defining fields for a specific view is something that I believe is good to have in the view or the controller, but not in the model
